File: app/src/main/java/com/github/oryanmat/trellowidget/activity/GeneralPreferenceFragment.java
package com.github.oryanmat.trellowidget.activity;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.ListPreference;
import android.preference.PreferenceFragment;

import com.github.oryanmat.trellowidget.R;
import com.github.oryanmat.trellowidget.util.ColorPreference;
import com.github.oryanmat.trellowidget.widget.TrelloWidgetProvider;

public class GeneralPreferenceFragment extends PreferenceFragment {
    static final String COLOR_FORMAT = "#%08X";

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        addPreferencesFromResource(R.xml.pref_general);

        SharedPreferences preferences = getPreferenceScreen().getSharedPreferences();
        listener.onSharedPreferenceChanged(preferences, getString(R.string.pref_text_size_key));
        listener.onSharedPreferenceChanged(preferences, getString(R.string.pref_back_color_key));
        listener.onSharedPreferenceChanged(preferences, getString(R.string.pref_fore_color_key));
        listener.onSharedPreferenceChanged(preferences, getString(R.string.pref_update_interval_key));
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceScreen().getSharedPreferences()
                .registerOnSharedPreferenceChangeListener(listener);
    }

    @Override
    public void onPause() {
        super.onPause();
        getPreferenceScreen().getSharedPreferences()
                .unregisterOnSharedPreferenceChangeListener(listener);
        TrelloWidgetProvider.updateWidgets(getActivity());
        TrelloWidgetProvider.updateWidgetsData(getActivity());
    }

    SharedPreferences.OnSharedPreferenceChangeListener listener =
            new SharedPreferences.OnSharedPreferenceChangeListener() {
                public void onSharedPreferenceChanged(SharedPreferences prefs, String key) {
                    setPreferenceChanges(key);
                }
            };

    private void setPreferenceChanges(String key) {
        if (key.equals(getString(R.string.pref_update_interval_key))) {
            ListPreference preference = (ListPreference) findPreference(key);
            int index = preference.findIndexOfValue(preference.getValue());
            preference.setSummary(String.format(getActivity()
                    .getString(R.string.pref_update_interval_value_desc), preference.getEntries()[index]));
        } else if (key.equals(getString(R.string.pref_text_size_key))) {
            ListPreference preference = (ListPreference) findPreference(key);
            int index = preference.findIndexOfValue(preference.getValue());
            preference.setSummary(preference.getEntries()[index]);
        } else if (key.equals(getString(R.string.pref_back_color_key))) {
            ColorPreference preference = (ColorPreference) findPreference(key);
            preference.setSummary(String.format(COLOR_FORMAT, preference.getColor()));
        } else if (key.equals(getString(R.string.pref_fore_color_key))) {
            ColorPreference preference = (ColorPreference) findPreference(key);
            preference.setSummary(String.format(COLOR_FORMAT, preference.getColor()));
        }
    }
}

File: app/src/main/kotlin/com/github/oryanmat/trellowidget/activity/GeneralPreferenceFragment.kt
package com.github.oryanmat.trellowidget.activity

import android.content.SharedPreferences
import android.os.Bundle
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.SwitchPreference
import androidx.preference.PreferenceFragmentCompat
import androidx.annotation.StringRes
import com.github.oryanmat.trellowidget.R
import com.rarepebble.colorpicker.ColorPreference
import com.github.oryanmat.trellowidget.widget.updateWidgets
import com.github.oryanmat.trellowidget.widget.updateWidgetsData
import com.github.oryanmat.trellowidget.util.Constants.T_WIDGET_TAG
import android.util.Log

const val COLOR_FORMAT = "#%08X"

class GeneralPreferenceFragment : PreferenceFragmentCompat() {
    private val listener = SharedPreferences
        .OnSharedPreferenceChangeListener { _, key ->
            try {
                setPreferenceChanges(key)
            } catch (e: NullPointerException) {
                Log.e(T_WIDGET_TAG, "Can't find corresponding preference to key $key\n${e.stackTraceToString()}")
            }
        }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val preferences = preferenceScreen.sharedPreferences
        listener.onSharedPreferenceChanged(preferences, getString(R.string.pref_text_size_key))
        listener.onSharedPreferenceChanged(preferences, getString(R.string.pref_back_color_key))
        listener.onSharedPreferenceChanged(preferences, getString(R.string.pref_fore_color_key))
        listener.onSharedPreferenceChanged(preferences, getString(R.string.pref_title_back_color_key))
        listener.onSharedPreferenceChanged(preferences, getString(R.string.pref_title_fore_color_key))
        listener.onSharedPreferenceChanged(preferences, getString(R.string.pref_title_use_unique_color_key))
        listener.onSharedPreferenceChanged(preferences, getString(R.string.pref_update_interval_key))
        listener.onSharedPreferenceChanged(preferences, getString(R.string.pref_display_board_name_key))
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.pref_general, rootKey)
    }

    override fun onDisplayPreferenceDialog(preference: Preference) {
        if (preference is ColorPreference)
            preference.showDialog(this, 0)
        else
            super.onDisplayPreferenceDialog(preference)
    }

    override fun onResume() {
        super.onResume()
        preferenceScreen.sharedPreferences?.registerOnSharedPreferenceChangeListener(listener)
    }

    override fun onPause() {
        super.onPause()
        preferenceScreen.sharedPreferences?.unregisterOnSharedPreferenceChangeListener(listener)
        requireActivity().updateWidgets()
        requireActivity().updateWidgetsData()
    }

    private fun setPreferenceChanges(key: String) {
        when (key) {
            getString(R.string.pref_update_interval_key) -> {
                val preference = findPreference<ListPreference>(key)!!
                val index = preference.findIndexOfValue(preference.value)
                preference.summary = String.format(
                    getString(
                        R.string.pref_update_interval_value_desc
                    ), preference.entries[index]
                )
            }

            getString(R.string.pref_text_size_key) -> {
                val preference = findPreference<ListPreference>(key)!!
                val index = preference.findIndexOfValue(preference.value)
                preference.summary = preference.entries[index]
            }

            getString(R.string.pref_back_color_key) -> {
                val preference = findPreference<ColorPreference>(key)!!
                preference.summary = String.format(COLOR_FORMAT, preference.color)
            }

            getString(R.string.pref_fore_color_key) -> {
                val preference = findPreference<ColorPreference>(key)!!
                preference.summary = String.format(COLOR_FORMAT, preference.color)
            }

            getString(R.string.pref_title_back_color_key) -> {
                val preference = findPreference<ColorPreference>(key)!!
                preference.summary = String.format(COLOR_FORMAT, preference.color)
            }

            getString(R.string.pref_title_fore_color_key) -> {
                val preference = findPreference<ColorPreference>(key)!!
                preference.summary = String.format(COLOR_FORMAT, preference.color)
            }

            getString(R.string.pref_title_use_unique_color_key) -> {
                val preference = findPreference<SwitchPreference>(key)!!
                with(preference) {
                    summary = getString(R.string.pref_title_use_unique_color_desc)
                    colorPreference(R.string.pref_title_fore_color_key).isEnabled = isChecked
                    colorPreference(R.string.pref_title_back_color_key).isEnabled = isChecked
                }
            }

            getString(R.string.pref_display_board_name_key) -> {
                val preference = findPreference<SwitchPreference>(key)!!
                preference.summary = getString(R.string.pref_display_board_name_desc)
            }
        }
    }

    private fun colorPreference(@StringRes key: Int) =
        findPreference<ColorPreference>(getString(key))!!
}